File: 2022/haskell/06/Main.hs
module Main (main) where

isUnique :: [Char] -> Bool
isUnique [] = True
isUnique (x:xs) = x `notElem` xs && isUnique xs

findMarker :: String -> Int -> Int -> Int
findMarker s pointer size = if isUnique (take size s)
    then pointer + size
    else findMarker (tail s) (pointer + 1) size

main :: IO()
main = do
    inputFile <- readFile "../../inputs/06.txt"
    let run = (\size -> findMarker inputFile 0 size)
    let result1 = run 4
    let result2 = run 14
    print result1
    print result2

File: 2022/python/06/main.py
result_1 = 0
result_2 = 0

def is_unique(subsignal):
    chars_seen = ''
    for c in subsignal:
        if c in chars_seen:
            return False
        chars_seen += c
    return True

def find_marker(signal, n):
    p1, p2 = 0, n
    while p2 < len(signal):
        subsignal = signal[p1:p2]
        if is_unique(subsignal):
            return p2
        p1, p2 = p1+1, p2+1

with open('../../inputs/06.txt', 'r') as input_file:
    signal = input_file.read()

result_1 = find_marker(signal, 4)
result_2 = find_marker(signal, 14)
print(result_1)
print(result_2)
